### PR TITLE
Don't throw when create HashSet from fresh HasSet source

### LINKED ISSUE #5130: System.Collections.Generic: HashSet constructor throwing when passed another HashSet
The constructor for HashSet that takes an IEnumerable is throwing an unexpected exception when passed another HashSet.

repro test added to System.Collections test proj:
```
[Fact]
public void CopyConstructor()
{
    HashSet<T> set = new HashSet<T>();
    HashSet<T> set2 = new HashSet<T>(set); // throws NullReferenceException
}
```

Stack trace:
```
    System.Collections.Tests.HashSet_Generic_Tests_int.CopyConstructor [FAIL]
       System.NullReferenceException : Object reference not set to an instance of an object.
       Stack Trace:
          D:\git\corefx\src\system.collections\src\System\Collections\Generic\HashSet.cs(118,0): at System.Collections.Generic.HashSet`1..ctor(IEnumerable`1 collection, IEqualityComparer`1 comparer)
          D:\git\corefx\src\system.collections\src\System\Collections\Generic\HashSet.cs(96,0): at System.Collections.Generic.HashSet`1..ctor(IEnumerable`1 collection)
          D:\git\corefx\src\system.collections\tests\Generic\HashSet\HashSet.Generic.Tests.cs(39,0): at System.Collections.Tests.HashSet_Generic_Tests`1.CopyConstructor()
```

The throwing code was introduced in #4349 by Toni Petrina and is snippeted here:
```
...
 var otherAsHashSet = collection as HashSet<T>;
            if (otherAsHashSet != null && AreEqualityComparersEqual(this, otherAsHashSet))
            {
                _buckets = (int[])otherAsHashSet._buckets.Clone(); // THROWS
                _slots = (Slot[])otherAsHashSet._slots.Clone();
...
```

Note that it doesn't fail when using the System.Collections.dll in 4.0.11-rc2-23616.

### PR DESCRIPTION
**Pull request by JonHanna**
Fixes #5130

### CODE REVIEW

File: src/System.Collections/src/System/Collections/Generic/HashSet.cs [hunk 1]
```diff
@@ -115,6 +115,11 @@ public HashSet(IEnumerable<T> collection, IEqualityComparer<T> comparer)
             var otherAsHashSet = collection as HashSet<T>;
             if (otherAsHashSet != null && AreEqualityComparersEqual(this, otherAsHashSet))
             {
+                if (otherAsHashSet._buckets == null)
+                {
+                    // Source hasn't initialized buckets yet, so neither need this.
+                    return;
+                }
                 _buckets = (int[])otherAsHashSet._buckets.Clone();
```
Comment (bartonjs) [inline src/System.Collections/src/System/Collections/Generic/HashSet.cs]: I'd have a `Debug.Assert(otherAsHashSet._slots != null)` here, personally.

**Comment by bartonjs** [comment]
LGTM

File: src/System.Collections/src/System/Collections/Generic/HashSet.cs (hunk 1 above)
Comment (JonHanna) [inline src/System.Collections/src/System/Collections/Generic/HashSet.cs]: Not a bad idea. Will add.

File: src/System.Collections/src/System/Collections/Generic/HashSet.cs [hunk 2]
```diff
@@ -115,13 +115,18 @@ public HashSet(IEnumerable<T> collection, IEqualityComparer<T> comparer)
...
+                    Debug.Assert(otherAsHashSet._slots == null);
+                    return;
+                }
                 _buckets = (int[])otherAsHashSet._buckets.Clone();
                 _slots = (Slot[])otherAsHashSet._slots.Clone();
 
                 _count = otherAsHashSet._count;
                 _lastIndex = otherAsHashSet._lastIndex;
                 _freeList = otherAsHashSet._freeList;
-                _version = otherAsHashSet._version;
```
Comment (JonHanna) [inline src/System.Collections/src/System/Collections/Generic/HashSet.cs]: Noticed this unnecessary assignment while working on this constructor. Might as well remove it here.

**Comment by JonHanna** [comment]
Test this please,  Innerloop Windows_NT Release Build and Test

**Comment by ianhays** [comment]
LGTM. Thanks, Jon.

**Comment by JonHanna** [comment]
Rolling everything in this into #5147, so closing this.

**Comment by GuardRex** [comment]
@JonHanna Shout out for the work you've been doing over the last several months. Just wanted to say 'thanks' from a fellow community cat. You ROCK!

**Comment by JonHanna** [comment]
@GuardRex gosh! I've just been having fun. These libraries being opened is like being chauffeured around in a car for over a decade and then suddenly being allowed to pop the bonnet and get my hands greasy :)

**Comment by GuardRex** [comment]
@JonHanna Me, too! More **FUN** than a barrel of ... of ... servers!

**Pull request closed without merging**